File: tree/traversal/level_traversal/rightside_view.go
/*
 * @Author: yangxcc
 * @version: 1.0
 * @Date: 2022-10-23 16:58:24
 * @LastEditTime: 2023-01-19 14:27:11
 */
package level_traversal

import "trouble_solver/tree"

// 所谓右视图，其实就是层序遍历每一层的最后一个元素
func rightSideView(root *tree.TreeNode) []int {
	if root == nil {
		return []int{}
	}

	ans := []int{}
	queue := []*tree.TreeNode{}
	queue = append(queue, root)
	for len(queue) > 0 {
		size := len(queue) - 1
		for size >= 0 {
			node := queue[0]
			queue = queue[1:]
			if size == 0 {
				ans = append(ans, node.Val)
			}
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
			size--
		}
	}
	return ans
}

File: tree/traversal/level_traversal/RightsideView.java
package tree.traversal.level_traversal;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

import tree.TreeNode;

/**
 * leetcode 199 middle 二叉树的右视图
 * 
 * 该题没什么难度，就是找到层序遍历的最后一个节点
 */
public class RightsideView {
    public List<Integer> rightSideView(TreeNode root) {
        if (root == null) {
            return new ArrayList<>();
        }

        List<Integer> ans = new ArrayList<>();
        Deque<TreeNode> q = new ArrayDeque<>();
        q.add(root);
        while (!q.isEmpty()) {
            int size = q.size();

            for (int i = 0; i < size; i++) {
                TreeNode node = q.poll();

                if (node.left != null) {
                    q.add(node.left);
                }
                if (node.right != null) {
                    q.add(node.right);
                }

                if (i == size - 1) {
                    ans.add(node.val);
                }
            }
        }

        return ans;
    }
}
